design-proposals/klip-19-materialize-views.md

```diff
@@ -0,0 +1,223 @@
+and can not be updated directly.
+
+The existing `CREATE TABLE AS SELECT` statement will be removed from KSQL syntax and reserved for future use.
+If/when reintroduced it may create a table snapshot, i.e. a `TABLE`, not a `VIEW`, that is initialized from the 
+supplied query, i.e. a one-time initialization, not a continuous query/update. Though such functionality may 
+be better expressed using SQL `INSERT INTO` syntax. Alternatively, `CTAS` could be reintroduced as a continuously 
+updating table if we were to also offer correct `INSERT`, and later `UPDATE` and `DELETE` semantics, which may be possible.
+
+In case you spotted the edge-case, let's discuss it: what about the case where the data in Kafka is the _source dataset_, 
+but its managed by another system? As we know, KSQL can not offer correct `INSERT` table semantics, but it's still a 
+_source_, not a _derived dataset_, right? So how does this fit in the `TABLE` vs `VIEW` world? Well, KSQL may be reading 
+from the _source dataset_, but its essentially making a copy when it reads the data. That copy is a `VIEW`. Hence, users
+should import such a data set using `CREATE MATERIALIZED VIEW`.
+
+### What about streams? 
+
+Streams don't have the complexities of per-key update semantics that tables have. So, it's OK to insert into them, right?
+
+Well, streams are certainly not as clear cut as tables. But maybe there's a case for differentiating between a stream and
+a read-only stream view. 
+
+Consider a stream coming from some upstream system, e.g a click stream from a web-tier. Should a KSQL user be able to 
+add rows to such a stream? Probably not, but the true answer is most likely 'it depends'. It depends on how the organisation
+if using KSQL and modeling data in Kafka. Its possible the clickstream should be seen as read-only to KSQL users as its 
+owned by the web tier that produces it, and its maybe used by other systems too. On the flip side, its also possible that
+appending new rows to the stream is something they want to allow in KSQL.
+
+It can be argued that using ACLs are a good fit for controlling if a KSQL user can insert into a stream. There's nothing
+semantically wrong with allow the insert: a stream is just an append-only list of immutable 'facts'. Appending one more 
+fact does not change the previous or following facts. However, supporting read-only stream views may provide an easier 
+mental model for users to work with. If a user does a `CREATE MATERIALIZED STREAM VIEW AS SELECT` rather than  
+`CREATE STREAM AS SELECT` then it will be clear KSQL does not allow `INSERT`s into the stream.
+
+As a discussion point, and as a mirror of the proposed `MATERIALIZED VIEW` changes above, we propose introducing a 
+`CREATE MATERIALIZED STREAM VIEW` statement to allow users to import stream datasets as read-only.
+
+The existing `CREATE STREAM` statement will be for defining new streams or importing an existing topic as a stream, not a view.  
```

Comment:
Will this support the user creating a stream in KSQL that is both a) Owned by KSQL and b) Allows inserts?

Reply:
Yes. This is explained later in the doc.